client/dom/use-object-url.tsx

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from 'react'
+import { useValueAsRef } from '../state-hooks'
+
+/**
+ * A hook which creates an object URL out of the received file, while also making sure that stale
+ * object URLs are revoked.
+ *
+ * If the received argument is already a string, this hook is basically a noop which just returns
+ * the received string.
+ */
+export function useObjectUrl(file?: Blob | string): string | undefined {
+  const [objectUrl, setObjectUrl] = useState<string>()
+  const objectUrlRef = useValueAsRef(objectUrl)
```

Comment:
I would vastly prefer you just set the ref at the same time you set the state with something like this, otherwise you're depending on a render having run 100% of the time in order to revoke things properly (which is not guaranteed). Given that you're just using the state to trigger re-renders, it might be better to just *only* store the URL in a ref and use the stateless rerender hook we already have (to prevent inconsistency)